core/client/src/main/java/alluxio/client/block/stream/NettyPacketReader.java

```diff
@@ -0,0 +1,400 @@
+        } catch (IOException e) {
+          LOG.warn("Failed to close the NettyBlockReader (block: {}, address: {}).",
+              mId, mAddress, e);
+          try {
+            mChannel.close().sync();
+          } catch (InterruptedException ee) {
+            throw Throwables.propagate(ee);
+          }
+          return;
+        }
+      }
+    } finally {
+      if (mChannel.isOpen()) {
+        Preconditions.checkState(mChannel.pipeline().last() instanceof Handler);
+        mChannel.pipeline().removeLast();
+        resume();
+      }
+      BlockStoreContext.releaseNettyChannel(mAddress, mChannel);
+      mClosed = true;
+    }
+  }
+
+  /**
+   * @return bytes remaining
+   */
+  private long remaining() {
+    return mStart + mBytesToRead - mPosToRead;
+  }
+
+  /**
+   * @return true if there are too many packets pending
+   */
+  private boolean tooManyPacketsPending() {
+    return mPackets.size() >= MAX_PACKETS_IN_FLIGHT;
+  }
+
+  /**
+   * Add {@link Handler} to the channel pipeline.
+   */
+  private void addHandler() {
+    ChannelPipeline pipeline = mChannel.pipeline();
+    if (!(pipeline.last() instanceof RPCMessageDecoder)) {
+      throw new RuntimeException(String.format("Channel pipeline has unexpected handlers %s.",
+          pipeline.last().getClass().getCanonicalName()));
+    }
+    mChannel.pipeline().addLast(new Handler());
+  }
+
+  /**
+   * The netty handler that reads packets from the channel.
+   */
+  private class Handler extends ChannelInboundHandlerAdapter {
+    /**
+     * Default constructor.
+     */
+    public Handler() {}
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOException {
+      Preconditions.checkState(acceptMessage(msg), "Incorrect response type %s, %s.",
+          msg.getClass().getCanonicalName(), msg);
+
+      RPCProtoMessage response = (RPCProtoMessage) msg;
+      Protocol.Status status = ((Protocol.Response) response.getMessage()).getStatus();
+      if (!Status.isOk(status)) {
+        throw new IOException(String
+            .format("Failed to read block %d from %s with status %s.", mId, mAddress,
+                status.toString()));
+      }
+      mLock.lock();
+      try {
+        Preconditions.checkState(mPacketReaderException == null);
+        DataBuffer dataBuffer = response.getPayloadDataBuffer();
+        ByteBuf buf;
+        if (dataBuffer == null) {
+          buf = ctx.alloc().buffer(0, 0);
+        } else {
+          Preconditions.checkState(dataBuffer.getLength() > 0);
+          assert dataBuffer.getNettyOutput() instanceof ByteBuf;
+          buf = (ByteBuf) dataBuffer.getNettyOutput();
+        }
+        Preconditions.checkState(mPackets.offer(buf));
+        mNotEmptyOrFail.signal();
+
+        if (tooManyPacketsPending()) {
+          pause();
+        }
+      } catch (Throwable e) {
+        if (response.getPayloadDataBuffer() != null) {
+          response.getPayloadDataBuffer().release();
```

Comment:
This is to destroy the data to make sure that we don't leak ByteBuf.


 "exceptionCaught" passes the error to the reader. It will notify mNotEmptyOrFail.  Once that is notified, all subsequent NettyPacketReader#read throws IOException.